### ISSUE TITLE
SigV4Auth class does not construct canonical query string according to V4 spec

### ISSUE THREAD

**Issue by houglum** [boto/botocore#1951]
According to the docs at https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html, you're supposed to encode query params before sorting them:

> CanonicalQueryString specifies the URI-encoded query string parameters. You URI-encode name and values individually. You must also sort the parameters in the canonical query string alphabetically by key name. **The sorting occurs after encoding.**

However, the code here iterates over the sorted params and performs encoding **afterward**, inside of the loop for each pair:

https://github.com/boto/botocore/blob/526f1e0322bf87e74f69c894245ce8506b86e597/botocore/auth.py#L213-L220

As an FYI, I haven't checked other libraries, but the Java SDK one seems to not have this problem because it inserts encoded entries into a (inherently-sorted) `TreeMap` as it iterates over the originals.

**Comment by swetashre** [comment]
Thank you for your post. Are you having issues because of the code ? If so, can you provide a code sample and log ?

**Comment by houglum** [comment]
I don't think there exist any query parameters in the S3 API **today** where this could cause a problem.  Since query parameter *names* generally never contain unicode characters, the only problematic scenario I can imagine is one where:
- The API adds a param name that you can specify multiple times with different values (a generic example of this in an API is something like `api.someservice.com/v1/resources?filter=value1&filter=value2`)
- The values are something that could contain unicode characters (e.g. the name of a key/object in S3)
- And two different string values would sort differently when sorted as unicode code point values vs as their URI-encoded (ASCII) byte equivalents.

And to answer your question, no, but users use these libraries as reference implementations, so the libraries should really be consistent with each other and follow the algorithm defined in your docs / sigv4 spec.

**Comment by swetashre** [comment]
That is a complicated part of code and we are not going to change the working code unless someone is getting issue with the code.

**Comment by houglum** [comment]
I disagree that this is complicated; you either sort first and then encode, or encode first and then sort.  The latter order is correct, as it's explicitly described in S3's Signature Version 4 spec.  And if this code really were that complicated to the point where no one feels comfortable changing it without risking breaking it, I think that's a strong indication that more thorough testing is needed.

**Comment by houglum** [comment]
I see that the comment above about the code being complicated has been deleted, so I assume someone agrees with my points above :)

I came back to this issue because I was discussing with someone today whether query params should be sorted first or encoded first.  I assumed that this would have been fixed by now and hoped to point them to this issue, but it looks like this has just been sitting here untouched for 5 months even after I've offered a PR that fixes it.  I'm unsure as to why the fix hasn't been accepted; it seems like making your code follow the spec described in your docs would be a win for everyone.
